### DISCUSSION TITLE
Adding some server-side validation

### DISCUSSION THREAD

**Discussion by bleepsandblops** [solspace/craft-freeform#370] [Questions & Answers] [answered]
I'm trying to add some simple server-side validation (testing if email is in a certain format). I've seen this event:
https://docs.solspace.com/craft/freeform/v3/developer/form-events.html#on-validating-the-form

But it's not clear to me how to access the submission data from it.

Thank you

**Comment by gustavs-gutmanis** [comment, marked as answer]
Hi @bleepsandblops 

I would suggest using the [field valiadation event](https://docs.solspace.com/craft/freeform/v3/developer/field-events.html#before-a-field-is-validated) instead.

Here's an example of how this might look in your specific case.
Note that the fields contain the posted values already in this step, so you access the submission data by asking for the field value.

```php
Event::on(
  FieldsService::class,
  FieldsService::EVENT_BEFORE_VALIDATE,
  function (ValidateEvent $event) {
    $form = $event->getForm();
    $field = $event->getField();

    // Only perform this validation when the field is in a specific form (optional)
    if ($field->getHandle() !== 'my-specific-form') {
        return;
    }

    // Only perform this validation on the specific field that you want
    if ($field->getHandle() !== 'my-email-field') {
      return;
    }

    // Example validation, this is where you write your logic
    // If the field doesn't pass the validation check, use the ::addError() method to attach errors to it
    if ($field->getValue() !== 'specific@email.com') {
        $field->addError("This requires a specific email");
    }
  }
)
```
